**Artificial/Deep Neural Network**



In [2]:
import numpy as np 
import pandas as pd
import tensorflow as tf
tf.__version__

'2.12.0'

In [3]:
dataset = pd.read_csv('/content/Churn_Modelling.csv')
dataset.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [4]:
# split the data into dependent and independent variable
x = dataset.iloc[:,3:-1]
y = dataset.iloc[:,-1]

In [5]:
x.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

In [18]:
x.shape

(10000, 11)

In [6]:
y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [8]:
x = pd.get_dummies(x, columns=['Geography', 'Gender'])
x


CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             619   42       2       0.00              1          1   
1             608   41       1   83807.86              1          0   
2             502   42       8  159660.80              3          1   
3             699   39       1       0.00              2          0   
4             850   43       2  125510.82              1          1   
...           ...  ...     ...        ...            ...        ...   
9995          771   39       5       0.00              2          1   
9996          516   35      10   57369.61              1          1   
9997          709   36       7       0.00              1          0   
9998          772   42       3   75075.31              2          1   
9999          792   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Geography_France  Geography_Germany  \
0                  1        101348.88                 1                  0   
1                  1        112542.58                 0                  0   
2                  0        113931.57                 1                  0   
3                  0         93826.63                 1                  0   
4                  1         79084.10                 0                  0   
...              ...              ...               ...                ...   
9995               0         96270.64                 1                  0   
9996               1        101699.77                 1                  0   
9997               1         42085.58                 1                  0   
9998               0         92888.52                 0                  1   
9999               0         38190.78                 1                  0   

      Geography_Spain  Gender_Female  Gender_Male  
0                   0              1            0  
1                   1              1            0  
2                   0              1            0  
3                   0              1            0  
4                   1              1            0  
...               ...            ...          ...  
9995                0              0            1  
9996                0              0            1  
9997                0              1            0  
9998                0              0            1  
9999                0              1            0  

[10000 rows x 13 columns]

In [9]:
x = x.drop(['Geography_France','Gender_Female'], axis = 1)

In [11]:
x.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42       2       0.00              1          1   
1          608   41       1   83807.86              1          0   
2          502   42       8  159660.80              3          1   
3          699   39       1       0.00              2          0   
4          850   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
0               1        101348.88                  0                0   
1               1        112542.58                  0                1   
2               0        113931.57                  0                0   
3               0         93826.63                  0                0   
4               1         79084.10                  0                1   

   Gender_Male  
0            0  
1            0  
2            0  
3            0  
4            0

In [14]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
x_stand = sc.fit_transform(x)
x_stand

array([[-0.32622142,  0.29351742, -1.04175968, ..., -0.57873591,
        -0.57380915, -1.09598752],
       [-0.44003595,  0.19816383, -1.38753759, ..., -0.57873591,
         1.74273971, -1.09598752],
       [-1.53679418,  0.29351742,  1.03290776, ..., -0.57873591,
        -0.57380915, -1.09598752],
       ...,
       [ 0.60498839, -0.27860412,  0.68712986, ..., -0.57873591,
        -0.57380915, -1.09598752],
       [ 1.25683526,  0.29351742, -0.69598177, ...,  1.72790383,
        -0.57380915,  0.91241915],
       [ 1.46377078, -1.04143285, -0.35020386, ..., -0.57873591,
        -0.57380915, -1.09598752]])

In [16]:
# split the data into training and testing
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x_stand, y, test_size=0.2, random_state=1)

# Building Deep Neural Network (MultiLayerPerceptron)

In [17]:
from tensorflow import keras

In [19]:
# creat APT
dnn = tf.keras.models.Sequential()
#Adding first Hidden Layer
dnn.add(tf.keras.layers.Dense(units=6, activation='relu'))           # in 1st hidden layer i am giving 6 neuron

#Adding second Hidden Layer
dnn.add(tf.keras.layers.Dense(units=6, activation='relu'))           # units should be equal or less than hidden layer 1

# Tensorflow is a backend language, keras is a frontend language, layers means hidden layer
# Dense means fully connected ie., each input data is individually connected to each layers/neuron and we are using relu as a activation functions

# adding output
dnn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))         # if it is binary class classification we can use sigmoid, if it is multiclass classification can use softMax because it gives multipule probability values

# Train the model
dnn.compile(optimizer='adam', loss = 'binary_crossentropy', metrics = ['accuracy'])   # since it is a binary classification we use binary crossentropy and we evaluate using accuracy mostly

# apply training dataset and fit the model
dnn.fit(x_train, y_train, batch_size = 32, epochs=100)

Epoch 1/100
250/250 [==============================] - 7s 3ms/step - loss: 0.5558 - accuracy: 0.7940
Epoch 2/100
250/250 [==============================] - 1s 3ms/step - loss: 0.4713 - accuracy: 0.7972
Epoch 3/100
250/250 [==============================] - 1s 3ms/step - loss: 0.4316 - accuracy: 0.8048
Epoch 4/100
250/250 [==============================] - 1s 3ms/step - loss: 0.4160 - accuracy: 0.8215
Epoch 5/100
250/250 [==============================] - 1s 3ms/step - loss: 0.4083 - accuracy: 0.8223
Epoch 6/100
250/250 [==============================] - 1s 3ms/step - loss: 0.4023 - accuracy: 0.8248
Epoch 7/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3976 - accuracy: 0.8241
Epoch 8/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3934 - accuracy: 0.8278
Epoch 9/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3892 - accuracy: 0.8278
Epoch 10/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3856 - accura

In [20]:
dnn.fit(x_train, y_train, batch_size = 32, epochs=100)

Epoch 1/100
250/250 [==============================] - 1s 4ms/step - loss: 0.3360 - accuracy: 0.8597
Epoch 2/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3358 - accuracy: 0.8606
Epoch 3/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3358 - accuracy: 0.8593
Epoch 4/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3357 - accuracy: 0.8605
Epoch 5/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3353 - accuracy: 0.8615
Epoch 6/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3359 - accuracy: 0.8608
Epoch 7/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3352 - accuracy: 0.8605
Epoch 8/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3348 - accuracy: 0.8590
Epoch 9/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3354 - accuracy: 0.8600
Epoch 10/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3353 - accura

In [21]:
# Predict the test result
y_pred = dnn.predict(x_test)

63/63 [==============================] - 0s 2ms/step


In [23]:
y_pred = (y_pred>0.5)

In [24]:
from sklearn.metrics import accuracy_score, classification_report

In [25]:
accuracy_score(y_test, y_pred)

0.864

In [27]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.89      0.95      0.92      1585
           1       0.74      0.53      0.62       415

    accuracy                           0.86      2000
   macro avg       0.81      0.74      0.77      2000
weighted avg       0.86      0.86      0.86      2000



In this employee churn modeling project, we developed a deep learning model to predict employee churn within an organization. The model achieved an accuracy of 86% on both the training and testing datasets, indicating a balanced and reliable performance. This means that the model has successfully learned the patterns and features that distinguish between employees who are likely to stay and those who are likely to leave the organization.

The high accuracy achieved by the model suggests its potential as a valuable tool for identifying employees at risk of churn. By proactively identifying such employees, organizations can take appropriate actions to mitigate churn, such as implementing retention strategies, providing additional support, or addressing any underlying concerns.

Overall, this employee churn classification model demonstrates the effectiveness of deep learning in predicting employee churn and provides valuable insights for organizations to make data-driven decisions in their employee retention efforts.

